packages/runtime/src/components/hooks/usePageSnippets.ts

```diff
@@ -0,0 +1,106 @@
+import { useEffect, useMemo, useRef, useState } from 'react'
+import { useIsInBuilder } from '../../react'
+import { MakeswiftPageDocument } from '../../next'
+import deepEqual from '../../utils/deepEqual'
+import { useMakeswiftHostApiClient } from '../../next/context/makeswift-host-api-client'
+import { useSyncExternalStore } from 'use-sync-external-store/shim'
+import { Page as PageType } from '../../api'
+
+const SnippetLocation = {
+  Body: 'BODY',
+  Head: 'HEAD',
+} as const
+
+export type SnippetLocation = typeof SnippetLocation[keyof typeof SnippetLocation]
+
+export type Snippet = {
+  builderEnabled: boolean
+  cleanup: string | null
+  code: string
+  id: string
+  liveEnabled: boolean
+  location: SnippetLocation
+}
+
+const filterUsedSnippetProperties = ({
+  code,
+  builderEnabled,
+  liveEnabled,
+  location,
+  cleanup,
+}: Snippet) => ({
+  code,
+  builderEnabled,
+  liveEnabled,
+  location,
+  cleanup,
+})
+
+export function usePageSnippets({ page }: { page: MakeswiftPageDocument }) {
+  const isInBuilder = useIsInBuilder()
+  const [snippets, setSnippets] = useState(page.snippets)
+
+  // We're using cached results here for page snippets and site fonts so that anytime the user
+  // changes the snippets or fonts on the builder, the change would be reflected here.
+  // See this PR for discussions and things we can do to improve it in the future:
+  // https://github.com/makeswift/makeswift/pull/77
+  const cachedPage = useCachedPage(isInBuilder ? page.id : null)
+  useEffect(() => {
+    if (cachedPage == null) return
+
+    const oldSnippets = snippets.map(filterUsedSnippetProperties)
+    const newSnippets = cachedPage.snippets.map(filterUsedSnippetProperties)
+
+    if (deepEqual(newSnippets, oldSnippets)) return
+
+    setSnippets(cachedPage.snippets)
+  }, [cachedPage])
+
+  const filteredSnippets = useMemo(
+    () => snippets.filter(snippet => (isInBuilder ? snippet.builderEnabled : snippet.liveEnabled)),
+    [snippets, isInBuilder],
+  )
+
+  const headSnippets = useMemo(
+    () => filteredSnippets.filter(snippet => snippet.location === SnippetLocation.Head),
+    [filteredSnippets],
+  )
+
+  const previousHeadSnippets = useRef<MakeswiftPageDocument['snippets'] | null>(null)
+
+  useEffect(() => {
+    const headSnippetsToCleanUp = (previousHeadSnippets.current ?? [])
+      .filter(previousSnippet => previousSnippet.cleanup != null)
+      .filter(previousSnippet => !headSnippets.some(snippet => previousSnippet.id === snippet.id))
+
+    headSnippetsToCleanUp.forEach(snippetToCleanUp => {
+      if (snippetToCleanUp.cleanup == null) return
+
+      const cleanUp = new Function(snippetToCleanUp.cleanup)
+
+      try {
+        cleanUp()
+      } catch {
+        // Ignore errors from user input.
+      }
+    })
+
+    previousHeadSnippets.current = headSnippets
+  }, [headSnippets])
```

Comment:
How will this interact with App Router and `useServerInsertedHTML`? I'm guessing that now that HTML will come from the server and React won't do much to track it, but if we're now running this effect as `headSnippets` changes on client-side navigation will the cleanup work well? Have we tested what happens when there's different snippets in the head and we client-side navigate between those pages?